design-proposals/003-credentials.md

```diff
@@ -0,0 +1,226 @@
+ * Apply credentials to connections with minimum impact on the existing implementation
+
+## User Story
+
+As a citizen user I would like to connect to 3rd party service that require me to authenticate using credentials. Credentials that I provide or that are received during the authorization with the 3rd party service would need to be stored in Syndesis so I can manage them. For example when connecting to Salesforce the I grant Syndesis rights to perform actions on my behalf by, and in order to do that I would like to choose "connect to Salesforce" in Syndesis authenticate against Salesforce to authorize that access and not be prompted again for authorization. Syndesis in turn would apply the credentials received in the authorization process with Salesforce to my Salesforce connection, and perform any management needed, such as refreshing the the received credentials. I would like to have the ability to manage credentials that are stored in Syndesis.
+
+My requirements are:
+
+ * Simplicity in authorization flow ("connect to Salesforce" button)
+ * Secure storage of authorization data (credentials)
+ * Ability to manage credentials: delete and update
+ * Ability to apply credentials to connections
+
+## Discussion 
+
+The idea of credential provider is to encapsulate managing specifics of 3rd party SaaS credentials.
+
+Currently credentials are represented by connection properties that are marked as secret. This proposes an abstraction over existing connection properties with the ability to implement specific functionality for getting credentials such as OAuth tokens from 3rd party SaaS solutions. As such existing connection properties and their persistence as k8s secrets would remain the same, the UI would also have the choice of either managing the connection properties in the same way as it is currently done or can call into credentials provider implementation, specific to each SaaS solution, that would *apply* credentials to connection properties on the backend.
+Each 3rd party SaaS can have slightly different OAuth flow which mandates the need for credential provider, a plug in, possibly shipped as a separate artifact and discovered at runtime, mounted at `/api/{version}/credentials/{connector}` to facilitate the OAuth flow and manage (i.e. *apply*) the resulting credentials.
+Considering that OAuth requires per-user and per-application secrets, there is a need for system (Syndesis owned) and user (citizen user owned) credentials. Melding of those would be also responsibility of credential provider.
+
+Management of user and system credentials is also performed by credential provider, for instance initial setup of system credentials or refresh of user credentials.
+
+Credential provider consists of:
+ * **handler** - to implement OAuth flow (REST)
+ * **applicator** - to apply credentials to connection properties
+ * **deployer** - to augment k8s deployment descriptor
+ * **manager** - to manage credentials
+
```

Comment:
Maybe we should leave out the implementation detail here for a _credential provider_ (as these subcomponents are not referenced later on in this spec again). So it MAY consist of these components but I would be perfectly fine if it would be implemented otherwise. We should not go into too many details with the proposal.